Config File Update for V6

emmahodcroft opened the pull request:
Lots of work here. As previously discussed, `config` file should override anything supplied by command-line (CL). 

**My Idea**: During workshop or after following update tutorial, user gets basic run working in CL  & now wants to add more features. They should not have to supply whole `config` file straight away (hard). Should be able to pick&mix just what they need to customise. The rest comes from CL options. Makes building a `config` _easy_ as can be done bit-by-bit. 
However, you can also just pass everything in in a `config`, too, if that's easier.

Here's a list of changes, but also observations on behaviour/checks that I did.

_These changes work towards this aim and fix other stuff I encountered on the way._

## View options

- v1 uses `default` for display values; v2 uses `default_display`. We should encourage users to use the correct v2 value in their configs (less confusing) but tolerate old v1 hangovers. So using `defaults` gives a warning but converts it to `default_display` in background. (So also now supports using `default_display` in `config`.)

- Checked behaviour when set `default`/`default_display` to nonsensical values. `color_by` fails quietly with no detectable error in `augur` or `auspice`. `geo_resolution`, `distance_measure`, and `map_triplicate` give quiet (console) errors in `auspice` and revert to their default. (_Note: setting `map_triplicate` to 'cake' seems to be taken as 'true'... Mostly harmless, but should this happen?_) **TO-DO**: Explain these options in `--help` - or somewhere. I have not attempted to make these work in CL. Should we?

- `title` can now be specified either in `config` or CL or both (`config` overrules). I added a check/error if neither was supplied, but we currently have a default value for `--title` ('Analysis') so this error never happens. **What behaviour do we want here?**

- Maintainer info can also now be specified both places - same as `title` above, and same error question as above - we have "" and "" default values if nothing passed in, so error never prints. Is this what we want?

- Old v1 `maintainer` format (which works if passed in via `config`) was like this:
```
"maintainer": [
  "Emma H, Richard N, Robert D, Jan A",
  "http://neherlab.org"
],
```
But new v2 `maintainers` format in final JSON is like this:
```
"maintainers": [
  {
    "href": "http://neherlab.org",
    "name": "Emma Hodcroft"
  },
  {
    "href": "http://neherlab.org",
    "name": "Richard Neher"
  },
  {
    "href": "https://ki.se/en/people/janalb",
    "name": "Robert Dyrdak"
  },
  {
    "href": "https://ki.se/en/people/janalb",
    "name": "Jan Albert"
  }
],
```
I figured we want a way to _easily_ support multi-maintainers in `config`, which can use `maintainers` to match the new schema. I don't know much about JSON rules but I made up my own version:
```
"maintainers": [
  ["Emma H", "http://neherlab.org"], 
  ["Richard N","http://neherlab.org"],
  ["Robert D","https://ki.se/en/people/janalb"],
  ["Jan A", "https://ki.se/en/people/janalb"]
],
```
**Is this ok? If not, let's change it but stick with concept.** (As in, let's please not make users write 'href' and 'name' 4 times for 4 maintainers.)




## `color_options`/`colorings`:

- In general, options passed in `config` override CL, like we want. But more detail on this below.

- Checked that 'authors' can be specified as a colorby, just like anything else

- v1 uses `color_options`; v2 uses `colorings`. As with `default`/`default_display` above, we should encourage users to use the correct v2 value in their configs (less confusing) but tolerate old v1 hangovers. So using `color_options` gives a warning but converts it to `colorings` in background. (So also now supports using `colorings` in `config`.)

### `process_colorings()`

- `process_colorings()` was overwriting any JSON-format info (`title`, `type` etc) passed in. Looks at type now and acts accordingly. 

- Default for anything (`config` or CL) without a `type` is `categorical` - checked later to see if could be `continuous` - should add other checks in future.

- `type` is checked to ensure is one of: `continuous`, `ordinal`, `categorical`, `boolean` in line with Schema. If is `discrete` (v1 term) a warning is printed and this is internally converted to `categorical`. If anything else, an error is printed and the trait is removed from the export. **Is this ok?**

- v1 uses `legendTitle` and `menuItem`; v2 uses `title`. If these are present, a warning is printed, but internally `legendTitle` or `menuItem` (in that order) is used for `title`, and these fields are removed to prevent them from being written out to the JSON.

- v1 value `key` is no longer used. I checked it's completely ignored and not written into JSON.

- Checked that genotype data passed in via `config` `gt` and date data passed in via `config` `num_date` (as was done in v1) is completely ignored. 

- Checked that you can pass in traits via `config` with no values (no `title` or `type`) and they will be handled just like as if passed in by CL.

## Color-by/Traits Discord Question
Currently, to get 'traits' (all data on nodes) we get from any of the following: anything passed in via `--node-data` (clades, `augur traits` output, AA/Nuc muts, branchlengths), anything passed in with `--geography-traits`, anything listed under `colorings` in a `config`, and anything passed in with `--extra-traits`. 

This is used to put data on the nodes to be written out and to set up filters. 

'Color-by's (things you can select from the drop-down in `auspice`) are a subset of 'traits' but can be decided separately. If using CL (no `config`) they are the same. If using a `config`, then they are only those things listed under `colorings`/`color_options` in the `config`.

This can lead to some slightly weird behaviour:
- If something is passed in in `--extra-traits` but is not in the `config`, it cannot be colored by (not in drop-down), but _will_ be on the nodes of the final JSON and _will_ be a filter option (unless the user manually sets the filter options). 

- The same is true (but perhaps even less obvious) if something comes in via `--node-data` from being run in `augur traits`. Ex, 'country' has been run through `augur traits`, but is _not_ specified as a color-by option in the `config` or in `--extra-traits`, and not given in `geo` (`config`) or `--geography-traits`, the information is still written in the final JSON (on nodes) and is still a filter option. (In short, you did not put country anywhere in export specs, but it's still 'there'.)

This seems odd to me. **How should we handle this?** If it's not given as a color-by in `config` (when using `config`) should it be excluded as a 'trait' as well? What if it's not given as a color-by but it is given as `geo`? (Then we need the node-data to map it - does this make sense without coloring-by as well?)

A similar situation is true for Clades. If passed in via `--node-data` from running in `augur clades`, but is NOT given in `config` color options, then it won't appear in the color-by drop-down, but it will: still have clades appear as branch labels, appear as a filter, be written on nodes in the JSON. This case was more clear to me so I handled it. Unlike 'traits', clade info is generated by a separate `augur` command and passed into `--node-data` as a separate file. So excluding it entirely is easy (just don't pass in the file), whereas to exclude a 'trait' is harder (the 'traits' file may have multiple traits, so may require re-running `augur traits`). 

So, I took an 'all-or-nothing' approach. If you pass in `clades` data via `--node-data`, then you will get it as a color-by option (as well as branch labels, filter, etc) whether it's specified or not. If not wanted, exclude file. **Is this ok?**


## BUGS:
- `continuous` traits are still made a filter option (unless user uses a `config` to explicitly declare what should be a filter). This is very silly:
![image](https://user-images.githubusercontent.com/14290674/61318018-64716780-a804-11e9-8085-de6868beded6.png)

- If I put into `config` or `--extra-traits` (if using CL only) a color-by that doesn't exist in the data, I get no errors from `export v2`, but this really messes up `auspice`:
![image](https://user-images.githubusercontent.com/14290674/61317503-625ad900-a803-11e9-8cc3-05171ecc3035.png)
This persists in the legend even after you try to change to a different color-by:
![image](https://user-images.githubusercontent.com/14290674/61317551-7999c680-a803-11e9-9829-e7e338ff81fd.png)

- If I put a `categorical` trait (something with string values) to `continuous`, it totally kills `auspice`:
![image](https://user-images.githubusercontent.com/14290674/61317741-eca33d00-a803-11e9-8cc4-08f2f7409ecf.png)
The inverse does no harm as it's picked up by our check and transformed back into `continuous`... But do we always want this? **Should we have a way to specify number-only data as categorical?** (I think probably yes.)


It would be super-good to have more people testing this out and seeing what they think of these options. I've put a how-to-switch guide in the docs (on a branch) if you need help!
All reviewers welcome - unsure who is currently working on `v6` stuff.

trvrb commented:
This is really fantastic work Emma. One thing: my expectation (or least surprise) would be for command line input to override config file. Thinking of the common use case where a user will have borrowed a config file from another build but want to override specific elements.

jameshadfield commented:
Nice one @emmahodcroft. I agree with anything I don't talk about below!

## View Options

> Explain these [config] options in --help - or somewhere

💯 these have been really poorly documented. Help might clutter things up tho. Perhaps a page on read-the-docs or nextstrain.org/docs and include the URL in the help message?

> we currently have a default value for --title ('Analysis') so this error never happens. What behaviour do we want here?

Get rid of the default. Auspice works fine without a `title` 😃 

> Maintainer info ... [has] "" and "" default values.

Allow JSONs without `maintainer`. Currently this crashes auspice but it shouldn't.

## `process_colorings()`

I've created changes in https://github.com/nextstrain/augur/pull/315 which will conflict, but I agree with all your points. I'll sort out the conflicts when that branch is merged. That's on me -- I'll try and merge both ASAP so we don't run into more confusion.

> Checked that genotype data passed in via config gt and date data passed in via config num_date (as was done in v1) is completely ignored.

In #315 I've allowed a "gt" key to be suplied by the config, as it allows the (rare I imagine) case where someone wants to call it something other than "genotype". Ditto for "num_date" with it's default title of "sampling date".

## Color-by discordance

> This seems odd to me. How should we handle this? If it's not given as a color-by in config (when using config) should it be excluded as a 'trait' as well? What if it's not given as a color-by but it is given as geo? (Then we need the node-data to map it - does this make sense without coloring-by as well?)

I don't know 😕 -- I made a comment in #315 to this effect! I think we want to merge the traits args, i.e. if the config defines `X` and the args define `Y` then you get `X & Y`. But I don't think this always holds for geo -- you can have a geo resolution that you don't want as a color-by.

## BUGS

> If I put into config or --extra-traits (if using CL only) a color-by that doesn't exist in the data, I get no errors from export v2, but this really messes up auspice

#315 fixes this 👍 

> Should we have a way to specify number-only data as categorical? 

Yes -- I think we only "guess" the type if it's not defined in the config. That's what #315 does.

rneher commented:
I also think command line should overwrite config.

emmahodcroft commented:
@trvrb & @rneher Oh man, I got this completely turned around in my head. My apologies. I knew we were going with with Tom suggested but managed to mix up what that was.

This overhauls a few things above. I'll think more today but my big question would be - so how will this work for allowing to specify exactly what color-by options are? We don't have any way of not including a trait as color-by in CL. If CL overrides `config`, then this means you cannot exclude a trait as being a color-by in `export`, as far as I can follow. (Or is this not technically overriding?)

(In a similar vein, we don't allow specification of `title` or `type` in CL, but we want them to be allowed in `config`. Perhaps this technically isn't overriding as it isn't a CL option, but it seems like it gives that appearance in execution? I think these kinds of things are how my thinking got flipped...)

In the meantime I will address @jameshadfield comments above that don't relate to this, but can flip `config`/CL priority once we've ironed it out (in a separate PR if that makes it easier for you to go ahead and merge w/315 your-tomorrow/my-tonight, James).

emmahodcroft commented:
> 💯 these have been really poorly documented. Help might clutter things up tho. Perhaps a page on read-the-docs or nextstrain.org/docs and include the URL in the help message?

Agree. Let's put them somewhere online, but let's not forget. Can we put a line about this above the help (or below the help, or somewhere in the help) with a placeholder URL "XXX" to help remind us to do this?

emmahodcroft commented:
@jameshadfield I've removed the defaults for `title` and `maintainers`/`url`. ~The error messages if they are missing remain for the moment, but these are easy to remove if we don't want them.~ So our Zika v2 test build has no title - this this error is causing it to fail! Ok, I'll remove the `--title` error :) 
Maintainer error remains for the moment but please remove we want.

I added the panel options for `--panels` to `--help` (was previously XXX). Checked that this works in CL, and changed so this now also works in `config`. 
I checked what happens if you pass in nonsense options - quietly ignored. User must supply at least 1 option. If you ask for 'frequencies' and you don't have the corresponding file, Auspice errors nicely. I assume this is true for other panels but didn't check. 
There is more panels logic in `export_v2.py`, but I didn't check it. 

I changed `--title` so users no longer have to use weird quotation combos when specifying the title in CL.

@james I will leave everything else until you can sort out with what you've already updated in #314, then I can pick back up!

jameshadfield commented:
Merging now -- i've added an issue about docs here: https://github.com/nextstrain/augur/issues/317 and we can address how exactly the CL overwrites the config file as we develop v6 🎉
